pkg/engine/params.go

```diff
@@ -221,11 +221,26 @@ func getParameters(cs *api.ContainerService, generatorCode string, aksEngineVers
 			addValue(parametersMap, "agentWindowsImageResourceGroup", properties.WindowsProfile.ImageRef.ResourceGroup)
 			addValue(parametersMap, "agentWindowsImageName", properties.WindowsProfile.ImageRef.Name)
 		} else {
-			addValue(parametersMap, "agentWindowsPublisher", properties.WindowsProfile.WindowsPublisher)
-			addValue(parametersMap, "agentWindowsOffer", properties.WindowsProfile.WindowsOffer)
-			addValue(parametersMap, "agentWindowsSku", properties.WindowsProfile.GetWindowsSku())
-			addValue(parametersMap, "agentWindowsVersion", properties.WindowsProfile.ImageVersion)
-
+			if properties.WindowsProfile.WindowsPublisher == "" {
+				addValue(parametersMap, "agentWindowsPublisher", cloudSpecConfig.OSImageConfig[api.WindowsDistro].ImagePublisher)
+			} else {
+				addValue(parametersMap, "agentWindowsPublisher", properties.WindowsProfile.WindowsPublisher)
+			}
+			if properties.WindowsProfile.WindowsOffer == "" {
+				addValue(parametersMap, "agentWindowsOffer", cloudSpecConfig.OSImageConfig[api.WindowsDistro].ImageOffer)
+			} else {
+				addValue(parametersMap, "agentWindowsOffer", properties.WindowsProfile.WindowsOffer)
+			}
+			if properties.WindowsProfile.WindowsSku == "" {
+				addValue(parametersMap, "agentWindowsSku", cloudSpecConfig.OSImageConfig[api.WindowsDistro].ImageSku)
+			} else {
+				addValue(parametersMap, "agentWindowsSku", properties.WindowsProfile.WindowsSku)
+			}
+			if properties.WindowsProfile.ImageVersion == "" {
+				addValue(parametersMap, "agentWindowsVersion", cloudSpecConfig.OSImageConfig[api.WindowsDistro].ImageVersion)
+			} else {
+				addValue(parametersMap, "agentWindowsVersion", properties.WindowsProfile.ImageVersion)
+			}
 		}
```

Comment:
Please add some unit tests to validate this logic

Reply:
Also, the properties.WindowsProfile properties here all get defaults in [https://github.com/Azure/aks-engine/blob/f9dd0e808e569787a7bc61119234fb1f403cef0c/pkg/api/defaults.go#L696-L720](https://github.com/Azure/aks-engine/blob/f9dd0e808e569787a7bc61119234fb1f403cef0c/pkg/api/defaults.go#L696-L720) so they should never be empty string.

---
In reply to: [364890111](https://github.com/Azure/aks-engine/pull/2548#discussion_r364890111) [](ancestors = 364890111)

Reply:
Sorry for missing that. I think that I can remove setWindowsProfileDefaults.

Reply:
Instead of removing setWindowsProfileDefaults, why not just keep the aks-engine code the same but remove these fields from the apimodel?

Reply:
If the goal here is to just use the aks-engine defaults i think that can already be accomplished by leaving WindowsPublisher, WIndowsOffer, WindowsSKU, and ImageVersion out of the WindowsProfile block in the cluster definition/apimodel.

Reply:
@marosset , you are right. If users want to use the default Windows OS image in aks-engine, they only need to remove the specified variables in my apimodel.
I want to use this PR for below purpose:
1. Keep consistent: Specify default image for Windows as Linux does
2. User can easily specify a Windows Image by Distro name.
Please see the update in my PR.

Reply:
Can the logic here be moved into GetWindowsPublisher(), GetWindowsOffer(), GetWindowsSku(), GetImageVersion() functions on WindowsProfile to keep things cleaner here?

Reply:
Update it.